deepinterpolation/generator_collection.py

```diff
@@ -897,112 +918,143 @@ class OphysGenerator(SequentialGenerator):
 
         # For backward compatibility
-        if "train_path" in self.json_data.keys():
-            self.raw_data_file = self.json_data["train_path"]
-        else:
-            self.raw_data_file = self.json_data["movie_path"]
-
+        self.cache_data = self.json_data.get("cache_data", False)
+        self.raw_data_file = self.json_data.get(
+            "train_path", self.json_data["movie_path"])
+        self.gpu_cache_full = self.json_data.get("gpu_cache_full", False)
+        self.normalize_cache = self.json_data.get("normalize_cache", False)
         self.batch_size = self.json_data["batch_size"]
-        movie_obj_point = h5py.File(self.raw_data_file, "r")
-        raw_data = movie_obj_point["data"]
+        self.movie_statistics_sample_size = \
+            self.json_data.get("movie_statistics_sample_size")
 
-        self.total_frame_per_movie = int(raw_data.shape[0])
+        self.total_frame_per_movie = int(self.movie_data.shape[0])
 
         self._update_end_frame(self.total_frame_per_movie)
         self._calculate_list_samples(self.total_frame_per_movie)
 
-        average_nb_samples = np.min([int(raw_data.shape[0]), 100])
-
-        # For backward compatibility
-        if "cache_data" in self.json_data.keys():
-            self.cache_data = self.json_data["cache_data"]
-        else:
-            self.cache_data = False
-
-        if self.cache_data:
-            print('Caching hdf5 file... \n')
-            if self.end_frame > 0:
-                self.raw_data = raw_data[0:np.min([self.total_frame_per_movie,
-                    self.end_frame+self.post_frame+self.pre_post_omission+1]), :, :]
-            else:
-                self.raw_data = raw_data[:, :, :]
-            local_data = self.raw_data[0:average_nb_samples, :, :].flatten()
-        else:
-            local_data = raw_data[0:average_nb_samples, :, :].flatten()
-            self.cache_data = False
-
+        average_nb_samples = np.min(
+            [int(self.movie_data.shape[0]), self.movie_statistics_sample_size])
+        local_data = self.movie_data[:average_nb_samples].flatten()
         local_data = local_data.astype("float32")
 
-        self.local_mean = np.mean(local_data)
-        self.local_std = np.std(local_data)
+        self.local_mean = local_data.mean()
+        self.local_std = local_data.std()
+
+        if self._gpu_available:
+            self.cached_index = None
+            self._data_tensor = None
+            if not self.gpu_cache_full:
+                # Get indices for partial movie cache
+                start = self.pre_frame + self.pre_post_omission
+                end = start + self.batch_size
+                self.batch_indices = list(range(start,end))
+                self.input_indices = np.vstack(
+                    [self.__index_generation(frame_index) \
+                        for frame_index in self.batch_indices])
+
+    @property
+    def movie_data(self) -> Union[np.ndarray, tf.Tensor]:
```

Comment:
why not convert to `Tensor` always and just work with that instead of 2 different data types?

Reply:
Actually, after testing this, the CPU processing when first converted to a tensor is much slower. Perhaps the slicing operation isn't nearly as efficient with tf.Tensor as opposed to np.array.

Reply:
ok. you mean that it is slow only the first time an operation is done on the tensor? I think that is a fine tradeoff to make since it simplifies the code and reduces the chance for bugs.

Reply:
Nope, it's slower for each batch that's being processed which consists of batch generation (slicing) and inference. My test on a movie truncated to 5000 frames indicated that it takes 36 minutes as a numpy array and 62 minutes as a tf.Tensor with CPU processing.

Reply:
That's bad. One of the reasons why I prefer pytorch.